man/errinjct.8
\."
\." Copyright (C) 2004 International Business Machines
\."
.TH ERRINJCT 8 "June 2007" Linux "Linux on Power Service Tools"
.SH NAME
errinjct \- inject errors into a system via RTAS.
.SH SYNOPSIS
.B /usr/local/sbin/errinjct
[ OPTIONS ]
.SH DESCRIPTION
The \fIerrinjct\fR tool enables RTAS to inject errors on a Power system.
This tool is for IBM internal use only. Currently, errinjct is only
supported on PowerVM systems.

.SH OPTIONS
The options to the \fIerrinjct\fR tool change depending on the functionality
requested.  Run \fIerrinjct\fR without any arguments to see a list of
supported functions.  See the explanations of the individual functions
below for more information.

.SH FUNCTIONS
.TP
\fBopen\fR
Open the RTAS error injection facility.
.TP
\fBclose\fR
Close the RTAS error injection facility
.TP
\fBcorrupted-dcache-start\fR
Start causing LI data cache errors
.TP
\fBcorrupted-dcache-end\fR
Stop causing LI data cache errors
.TP
\fBcorrupted-icache-start\fR
Start causing instruction cache errors
.TP
\fBcorrupted-icache-end\fR
Stop causing instruction cache errors
.TP
\fBcorrupted-slb\fR
Corrupt the SLB entry associated with a specific effective address
.TP
\fBioa-bus-error\fR
Simulate an error on the IOA bus
.TP
\fBplatform-specific\fR
Request the firmware perform a specific platform specific injection
.TP
\fBcorrupted-tlb-start\fR
Start corrupting TLB
.TP
\fBcorrupted-tlb-end\fR
Stop corrupting TLB

.SH open
Open the RTAS error injection facility

Usage: errinjct open [options]

Optional arguments:
  --dry-run      don't perform the action,
                 just print what would have been done
  -H --help      print usage information for a particular
                 function
  -v --verbose   be more verbose with messages
  -q --quiet     shhhh.... only report errors

.SH close
Close the RTAS error injection facility

Usage: errinjct close [optioons]

Mandatory argument:
  -k token       token returned from error inject open

Optional arguments:
  --dry-run      don't perform the action,
                 just print what would have been done
  -H --help      print usage information for a particular
                 function
  -v --verbose   be more verbose with messages
  -q --quiet     shhhh.... only report errors
  -C cpu         cpu to inject errors on

.SH corrupted-dcache-start / corrupted-dcache-end
Inject errors onto the L1 data cache.  Before using this option you
need to open the RTAS error injection facility via "errinjct open".
Afterwards pass this parameter in with the -k option and specify a
cpu to run on with the -C option.  Please note that you need to specify
the same cpu number to the dcache-start and dcache-end errinjct calls.

Usage: errinjct corrupted-dcache-start [options]
       errinjct dcache-start [options]

       errinjct corrupted-dcache-end [options]
       errinjct dcache-end [options]

Mandatory Arguments:
  -a action      type of D-cache error to inject
                     0: parity error
                     1: D-ERAT parity error
                     2: tag parity error
  -C cpu         cpu to inject errors on
  -k token       token returned from error inject open

Optional arguments:
  --dry-run      don't perform the action,
                 just print what would have been done
  -H --help      print usage information for a particular
                 function
  -v --verbose   be more verbose with messages
  -q --quiet     shhhh.... only report errors

Example:
  > errinjct open
  Error inject token = 2
  > errinjct corrupted-dcache-start -a 1 -k 2 -C 0
  .
  .
  .
  > errinjct corrupted-dcache-end -a 1 -k 2 -C 0
  > errinjct close -k 2

.SH corrupted-icache-start / corrupted-icache-end
Inject errors into the instruction cache.  Before using this option you
need to open the RTAS error injection facility via "errinjct open".
Afterwards pass this parameter in with the -k option and specify a
cpu to run on with the -C option.  Please note that you need to specify
the same cpu number to the icache-start and icache-end errinjct calls.

Usage: errinjct corrupted-icache-start [options]
       errinjct icache-start [options]

       errinjct corrupted-icache-end [options]
       errinjct icache-end [options]

Mandatory Arguments:
  -a action      type of I-cache error to inject
                     0: Parity error
                     1: I-ERAT partiy error
                     2: Cache directory 0 parity error
                     3: Cache directory 1 parity error
  -n nature      nature of I-cache error to inject
                     0: Single
                     1: Solid
                     2: Hang
  -C cpu         cpu to inject errors on
  -k token       token returned from error inject open

Optional arguments:
  --dry-run      don't perform the action,
                 just print what would have been done
  -H --help      print usage information for a particular
                 function
  -v --verbose   be more verbose with messages
  -q --quiet     shhhh.... only report errors

Example:
  > errinjct open
  Error inject token = 2
  > errinjct corrupted-icache-start -a 1 -n 0 -k 2 -C 0
  .
  .
  .
  > errinjct corrupted-icache-end -a 1 -n 0 -k 2 -C 0
  > errinjct close -k 2

.SH corrupted-slb
Corrupt the SLB entry associated with the specified effective address.

Usage: errinjct corrupted-slb [options]
       errinjct slb [options]

Mandatory Argument:
  -a addr        effective address associated with the
                 SLB entry to corrupt

Optional arguments:
  --dry-run      don't perform the action,
                 just print what would have been done
  -H --help      print usage information for a particular
                 function
  -v --verbose   be more verbose with messages
  -q --quiet     shhhh.... only report errors
  -C cpu         cpu to inject errors on
  -k token       token returned from error inject open

.SH ioa-bus-error
Inject errors onto an IO adapter (PCI) bus slot.

Usage: errinjct ioa-bus-error [options]
       errinjct eeh [options]

Mandatory arguments:
  -f function    IOA bus error to inject

  Specify a device either with the -s flag, the
  -p flag, or use the explicit BUID/config address flags.
  -s classpath   look up device by sysfs classpath
                 for example -s net/eth3
		 or -s scsi_host/host0

  -p loc-code    look up device by location code
                 for example -p "U0.1-P2-I1"

  Explicit BUID/config mandatory arguments:
  -c config_addr configure address of the IOA
  -h high_bits   high bits of PHB unit id
  -l lo_bits     lo bits of PHB unit id

Optional arguments:
  --dry-run      don't perform the action,
                 just print what would have been done
  -H --help      print usage information for a particular
                 function
  -v --verbose   be more verbose with messages
  -q --quiet     shhhh.... only report errors
  -a addr        address at which to report the error
  -m mask        address mask (defaults to 0x00000000)
  -C cpu         cpu to inject errors on
  -k token       token returned from error inject open

Functions for ioa-bus-error:
  0 - Load to PCI Memory Address Space
      inject an Address Parity Error
  1 - Load to PCI Memory Address Space
      inject a Data Parity Error
  2 - Load to PCI I/O Address Space
      inject an Address Parity Error
  3 - Load to PCI I/O Address Space
      inject a Data Parity Error
  4 - Load to PCI Configuration Space
      inject an Address Parity Error
  5 - Load to PCI Configuration Space
      inject a Data Parity Error
  6 - Store to PCI Memory Address Space
      inject an Address Parity Error
  7 - Store to PCI Memory Address Space
      inject a Data Parity Error
  8 - Store to PCI I/O Address Space
      inject an Address Parity Error
  9 - Store to PCI I/O Address Space
      inject a Data Parity Error
 10 - Store to PCI Configuration Space
      inject an Address Parity Error
 11 - Store to PCI Configuration Space
      inject a Data Parity Error
 12 - DMA read to PCI Memory Address Space
      inject an Address Parity Error
 13 - DMA read to PCI Memory Address Space
      inject a Data Parity Error
 14 - DMA read to PCI Memory Address Space
      inject a Master Abort Error
 15 - DMA read to PCI Memory Address Space
      inject a Target Abort Error
 16 - DMA write to PCI Memory Address Space
      inject an Address Parity Error
 17 - DMA write to PCI Memory Address Space
      inject a Data Parity Error
 18 - DMA write to PCI Memory Address Space
      inject a Master Abort Error
 19 - DMA write to PCI Memory Address Space
      inject a Target Abort Error

.SH platform-specific
Request the firmware perform a platform specific error injection.

Usage: errinjct platform-specific [options]

Mandatory argument:
  -f fname       file name to read platform specific
                 error injection data from
Optional arguments:
  --dry-run      don't perform the action,
                 just print what would have been done
  -H --help      print usage information for a particular
                 function
  -v --verbose   be more verbose with messages
  -q --quiet     shhhh.... only report errors
  -C cpu         cpu to inject errors on
  -k token       token returned from error inject open

.SH corrupted-tlb-start / corrupted-tlb-end
Inject errors into the TLB. Before using this option you
need to open the RTAS error injection facility via "errinjct open".
Afterwards pass this parameter in with the -k option and specify a
cpu to run on with the -C option.  Please note that you need to specify
the same cpu number to the tlb-start and tlb-end errinjct calls.

Usage: errinjct corrupted-tlb-start [options]
       errinjct tlb-start [options]

       errinjct corrupted-tlb-end [options]
       errinjct tlb-end [options]

Mandatory arguments:
  -C cpu         cpu to inject errors on
  -k token       token returned from error inject open

Optional arguments:
  --dry-run      don't perform the action,
                 just print what would have been done
  -H --help      print usage information for a particular
                 function
  -v --verbose   be more verbose with messages
  -q --quiet     shhhh.... only report errors

Example:
  > errinjct open
  Error inject token = 2
  > errinjct corrupted-tlb-start -k 2 -C 0
  .
  .
  .
  > errinjct corrupted-tlb-end -k 2 -C 0
  > errinjct close -k 2

.SH Authors
Written by Nathan Fontenot and Linas Vepstas


Changelog
powerpc-utils-1.2.27
=====================================================================
commit 8e22e19dab80a1d17ca5ad5cdf112148282223eb
Author: John Allen <jallen@linux.vnet.ibm.com>
Date:   Tue Sep 29 14:41:14 2015 -0400

    errinjct/rtas_dbg: Update Makefile
    
    This patch updates the Makefile to include errinjct and rtas_dbg.
    
    Signed-off-by: John Allen <jallen@linux.vnet.ibm.com>

commit ef5f204ca2d5de67ef47de1ac29169e7e39b54a6
Author: John Allen <jallen@linux.vnet.ibm.com>
Date:   Tue Sep 29 14:37:12 2015 -0400

    rtas_dbg: Initial contribution of the rtas_dbg tool
    
    This patch contributes the previously IBM internal rtas_dbg tool. The rtas_d
    tool enables rtas debug output to the system console for a given rtas call.
    This version of the code is functionally identical to the old internal versi
    but I have made some superficial changes to remove IBM confidential informat
    and improve coding style.
    
    Signed-off-by: John Allen <jallen@linux.vnet.ibm.com>

commit 784885efc522d49e29c3f1fb679a31e7ba9b413d
Author: John Allen <jallen@linux.vnet.ibm.com>
Date:   Tue Sep 29 14:35:14 2015 -0400

    errinjct: Initial contribution of errinjct tool
    
    This patch contributes the previously IBM internal errinjct tool. The errinj
    tool is used to inject artificial hardware errors into a system. This versio
    of the code is functionally identical to the old internal version, but I hav
    made some superficial changes to remove IBM confidential information and
    improve coding style.
    
    Changes in v2: Now checks for platform. If a user is not on a PowerVM system
    the tool exits and prints a descriptive error message. Updated the man page 
    document lack of support for non-PowerVM systems.
    
    Signed-off-by: John Allen <jallen@linux.vnet.ibm.com>

commit 16e49f436470d597bb29656ed02b40fcdf47fcc5
Author: Nathan Fontenot <nfont@linux.vnet.ibm.com>
Date:   Wed Sep 9 10:30:53 2015 -0400

    lparstat: print "Shared Physical CPUS in system" data
    
    The information for "Shared Physical CPUS in system" is not printed when
    the lparstat command is run. This patch adds this information to the list
    of items to print and modifies the name to match the output of the lparstat
    command on other OS'es.
    
    Signed-off-by: Nathan Fontenot <nfont@linux.vnet.ibm.com>

commit 4153cccdc21a0c8810aab169fb5b04a611c7b154
Author: Bharata B Rao <bharata@linux.vnet.ibm.com>
Date:   Tue Sep 8 11:38:35 2015 -0400

    drmgr: Don't return unsuable CPU from get_available_cpu()
    
    From: Bharata B Rao <bharata@linux.vnet.ibm.com>
    
    When querying CPU by name or index (and not count), let
    get_available_cpu() return NULL for CPU that is marked as unusable.
    This will prevent the callers (eg. add_cpus()) from infinitely
    querying for the same CPU.
    
    Signed-off-by: Bharata B Rao <bharata@linux.vnet.ibm.com>

commit d291c43e4ad1141591008e39e77549a6fd33d48a
Author: Nathan Fontenot <nfont@linux.vnet.ibm.com>
Date:   Tue Sep 8 11:35:09 2015 -0400

    drmgr: Generate error message when PHB is not found
    
    An error message should be generated when drmgr is asked to remove a PHB
    that does not exist. The current behavior is to just exit without any
    message.
    
    Signed-off-by: Nathan Fontenot <nfont@linux.vnet.ibm.com>

commit 8ff5144da07985cd84508465007f9031840d4ecd
Author: Vasant Hegde <hegdevasant@linux.vnet.ibm.com>
Date:   Tue Sep 8 11:33:24 2015 -0400

    update_flash: Stop kexec service before starting firmware update on PowerVM 
    
    We run 'reboot' call before initiating firmware update. If kexec service
    is enabled, then we endup booting into new kernel instead of starting
    firmware update.
    
    Commit 62dc6d38 fix this issue on PowerNV platform.. but missed to fix
    PowerVM LPAR.  This patch fixes above described issue on PowerVM LPAR.
    
    Signed-off-by: Vasant Hegde <hegdevasant@linux.vnet.ibm.com>

commit 7a94345ff14fbb6d18a885616a5a7f347daa95f1
Author: John Allen <jallen@linux.vnet.ibm.com>
Date:   Wed Sep 2 12:43:15 2015 -0400

    drmgr: Fix file leak in dr_lock/dr_unlock interaction
    
    This patch fixes a file leak in dr_lock(). The file descriptor
    is meant to be saved to a global variable dr_lock_fd, but we
    have an extraneous local variable named dr_lock_fd as well. So
    When we get to dr_unlock, the global file descriptor that it
    tries to close was never set and quietly fails.
    
    Signed-off-by: John Allen <jallen@linux.vnet.ibm.com>

commit 29b8f317985249912b4025f704aae16811ea1775
Author: Boqun Feng <boqun.feng@linux.vnet.ibm.com>
Date:   Wed Sep 2 12:19:54 2015 -0400

    drmgr: handle read failures in get_att_prop()
    
    In get_att_prop(), though rarely, reading from /proc or /sysfs may
    fail. This patch handles failures of these readings based on the fact,
    if we successfully read the property, both fread() and fscanf() will
    return 1. This patch also updates the fread() path so that we only
    try to read the size of the file and not the size of the buffer
    passed in.
    
    This can also fix the compiler WARNING saying that 'rc' is unused
    variable.
    
    Signed-off-by: Boqun Feng <boqun.feng@linux.vnet.ibm.com>
    Signed-off-by: Nathan Fontenot <nfont@linux.vnet.ibm.com>

commit 9b8a3b449de4508adca248631e7ffda62a9a4994
Author: John Allen <jallen@linux.vnet.ibm.com>
Date:   Wed Sep 2 12:14:09 2015 -0400

    drmgr, ppc64_cpu: Fix several memory/file leaks
    
    This patch fixes a few resource leak problems discoverd through
    static analysis.
    
    Signed-off-by: John Allen <jallen@linux.vnet.ibm.com>

commit 5fee10ac758e16d49e42d5cf0855df44b1f90a7a
Author: Nathan Fontenot <nfont@linux.vnet.ibm.com>
Date:   Wed Sep 2 11:03:28 2015 -0400

    drmgr: Combine the init_node() and examine_child() routines
    
    The complexity of the init_node() and examine_child() routines that call

src/errinjct/errinjct.c
/**
 * @file errinjct.c
 * @brief Hardware Error Injection Tool main
 * @author Nathan Fontenot <nfont@austin.ibm.com>
 * @author Linas Vepstas <linas@us.ibm.com>
 *
 * This program can be used to simulate hardware error events.
 * It uses librtas to inject artificial errors into the system.
 *
 * Copyright (c) 2004 IBM Corporation
 * Common Public License Version 1.0 (see COPYRIGHT)
 */

#define _GNU_SOURCE
#include <fcntl.h>
#include <getopt.h>
#include <stdlib.h>
#include <stdarg.h>
#include <string.h>
#include <unistd.h>
#include <sched.h>
#include <librtas.h>
#include <sys/stat.h>
#include "errinjct.h"
#include "pseries_platform.h"

#define EI_TOKEN_PROCFILE	"/proc/device-tree/rtas/ibm,errinjct-tokens"
#define EI_IBM_ERRINJCT		"/proc/device-tree/rtas/ibm,errinjct"

int verbose;
int dryrun;
int ei_token = -1;
int logical_cpu = -1;
int ext_help;
int be_quiet;
int debug;

/**
 * @var progname
 * @brief name this app was invoked as (argv[0])
 */
char *progname;

uint err_buf[EI_BUFSZ / sizeof(uint)];

/**
 * @var ei_funcs
 * @brief array of known RTAS error injection functionality
 */
static ei_function ei_funcs[] = {
	{
		.name = "open",
		.alt_name = NULL,
		.desc = "open the RTAS error injection facility",
		.rtas_token = -1,
		.arg = ei_open_arg,
		.func = ei_open
	},

	{
		.name = "close",
		.alt_name = NULL,
		.desc = "close the RTAS error injection facility",
		.rtas_token = -1,
		.arg = ei_close_arg,
		.func = ei_close
	},

	{
		.name = "corrupted-dcache-start",
		.alt_name = "dcache-start",
		.desc = "Start causing a LI data cache error",
		.rtas_token = -1,
		.arg = corrupted_dcache_arg,
		.func = corrupted_dcache
	},

	{
		.name = "corrupted-dcache-end",
		.alt_name = "dcache-end",
		.desc = "Stop causing a LI data cache error",
		.rtas_token = -1,
		.arg = corrupted_dcache_arg,
		.func = corrupted_dcache
	},

	{
		.name = "corrupted-icache-start",
		.alt_name = "icache-start",
		.desc = "Start causing an instruction cache error",
		.rtas_token = -1,
		.arg = corrupted_icache_arg,
		.func = corrupted_icache
	},

	{
		.name = "corrupted-icache-end",
		.alt_name = "icache-end",
		.desc = "Stop causing an instruction cache error",
		.rtas_token = -1,
		.arg = corrupted_icache_arg,
		.func = corrupted_icache
	},

	{
		.name = "corrupted-page",
		.alt_name = NULL,
		.desc = "Corrupt the specified location (and potentially surrounding locations up to the containing page)",
		.rtas_token = -1,
		.arg = NULL,
		.func = NULL
	},

	{
		.name = "corrupted-slb",
		.alt_name = "slb",
		.desc = "Corrupt the SLB entry associated with a specific effective address",
		.rtas_token = -1,
		.arg = corrupted_slb_arg,
		.func = corrupted_slb
	},

	{
		.name = "corrupted-tlb-start",
		.alt_name = "tlb-start",
		.desc = "Start corrupting TLB",
		.rtas_token = -1,
		.arg = corrupted_tlb_arg,
		.func = corrupted_tlb
	},

	{
		.name = "corrupted-tlb-end",
		.alt_name = "tlb-end",
		.desc = "Stop corrupting TLB",
		.rtas_token = -1,
		.arg = corrupted_tlb_arg,
		.func = corrupted_tlb
	},

	{
		.name = "fatal",
		.alt_name = NULL,
		.desc = "Simulate a platform fatal error",
		.rtas_token = -1,
		.arg = NULL,
		.func = NULL
	},

	{
		.name = "ioa-bus-error",
		.alt_name = "eeh",
		.desc = "Simulate an error on an IOA bus",
		.rtas_token = -1,
		.arg = ioa_bus_error_arg,
		.func = ioa_bus_error32
	},

	{
		.name = "ioa-bus-error-64",
		.alt_name = "eeh-64",
		.desc = "Simulate an error on a 64-bit IOA bus",
		.rtas_token = -1,
		.arg = ioa_bus_error_arg,
		.func = ioa_bus_error64
	},

	{
		.name = "platform-specific",
		.alt_name = "platform",
		.desc = "Request the firmware perform a platform specific error injection",
		.rtas_token = -1,
		.arg = platform_specific_arg,
		.func = platform_specific
	},

	{
		.name = "recovered-random-event",
		.alt_name = "random-event",
		.desc = "Simulate a recovered random event",
		.rtas_token = -1,
		.arg = NULL,
		.func = NULL
	},

	{
		.name = "recovered-special-event",
		.alt_name = "special-event",
		.desc = "Simulate a recoverd special (statistically significant) event",
		.rtas_token = -1,
		.arg = NULL,
		.func = NULL
	},

	{
		.name = "translator-failure",
		.alt_name = NULL,
		.desc = "Simulate a translator failure",
		.rtas_token = -1,
		.arg = NULL,
		.func = NULL
	},

	{
		.name = "upstream-IO-error",
		.alt_name = NULL,
		.desc = "Inject I/O error above the IOA",
		.rtas_token = -1,
		.arg = NULL,
		.func = NULL
	}
};

/**
 * perr
 * @brief Creates a formatted error message and prints it to stderr
 * @param fmt format of the message to be printed
 * @param ... variable args
 */
void perr(int error, const char *fmt, ...)
{
	char		buf[EI_BUFSZ];
	va_list		ap;
	int		len;

	memset(buf, 0, EI_BUFSZ);

	len = sprintf(buf, "%s: ", progname);
	va_start(ap, fmt);
	len += vsprintf(buf + len, fmt, ap);
	va_end(ap);

	if (error)
		len += sprintf(buf + len, ", %s\n", strerror(error));
	else
		len += sprintf(buf + len, "\n");

	buf[len] = '\0';

	fprintf(stderr, buf);
	fflush(stderr);
}

/**
 * ei_ext_usage
 * @brief Print the extended usage message for errinjct
 *
 * This will print the name and description for all currently supported
 * error injection functions.
 */
static void ei_ext_usage(void)
{
	int i;

	printf("Currently supported functions:\n");
	for (i = 0; i < NUM_ERRINJCT_FUNCS; i++) {
		if (ei_funcs[i].func != NULL)
			printf("    %-25s%s\n", ei_funcs[i].name,
			       ei_funcs[i].desc);
	}

	printf("\n");
	printf("Try \"%s function -H\" for more information\n", progname);
}

/**
 * print_optional_args
 * @brief Print the optional arguments to errijnjct
 *
 * There are several options to the errinjct tool that are not specific
 * to any of the error injection functions (i.e. global).  This routine
 * is provided for all of the error injection functions to use to print
 * these optional args.
 */
void print_optional_args(void)
{
	printf("Optional arguments:\n");
	printf(HELP_FMT, "--dry-run", "don't perform the action,");
	printf(HELP_FMT, "", "just print what would have been done");
	printf(HELP_FMT, "-H --help",
	       "print usage information for a particular function");
	printf(HELP_FMT, "-v --verbose", "be more verbose with messages");
	printf(HELP_FMT, "-vv", "turn on librtas tracing");
	printf(HELP_FMT, "-vvv", "turn on RTAS call argument tracing");
	printf(HELP_FMT, "-q --quiet", "shhhh.... only report errors");
}

/**
 * print_cpu_arg
 * @brief Print the usage for the "-C cpu" option
 *
 * Print the "-C cpu" part of a usage statement.  This argument may or
 * may not be mandatory (depending on the function) so we provide a
 * common print routine to centrlize messages.
 */
void print_cpu_arg(void)
{
	printf(HELP_FMT, "-C cpu", "cpu to inject errors on");
}

/**
 * check_cpu_arg
 * @brief Common routine to check if the "-C cpu" flag was used.
 *
 * This routine is provided to check if the "-C cpu" option was
 * specified and generate a common erro message.
 */
int check_cpu_arg(void)
{
	if (logical_cpu == -1) {
		perr(0, "Please specify a logical cpu with the -C option");
		return 1;
	}

	return 0;
}

/**
 * print_token_arg
 * @brief Print the "-k token" part of the usage statement
 *
 * Print the "-k token" part of the usage statement.  This argument may or
 * may not be mandatory (depending on the function) so we provide a
 * common print routine to centralize messages.
 */
void print_token_arg(void)
{
	printf(HELP_FMT, "-k token", "token returned from error inject open");
}

/**
 * check_token_arg
 * @brief Common routine to check if the "-k token" flag was used.
 *
 * This provides a common point for checking to see if the "-k token"
 * option was specified and generates a common error statement.
 */
int check_token_arg(void)
{
	if (ei_token == -1) {
		perr(0, "Please specify the error inject token with the -k option");
		return 1;
	}

	return 0;
}

/**
 * ei_usage
 * @brief Print the usage statement for errinjct
 */
static void ei_usage(void)
{
	printf("Usage: %s FUNCTION [OPTIONS]\n", progname);
	printf("This will inject an error into the system via rtas\n");

	ei_ext_usage();
}

/**
 * check_librtas_returns
 * @brief check the return code from librtas
 *
 * Check the return code from a call to librtas.  Librtas can return
 * values tat librtas specific values to indicate errors in librtas.
 * There is no method provided by librtas to do this so we have to do
 * it here.
 *
 * @param rc return code from a librtas call
 * @param ei_func functionality pointer
 */
void check_librtas_returns(int rc, ei_function *ei_func)
{
	switch (rc) {
	case -1: /* Hardware Error */
		perr(0, "RTAS: %s: Hardware error (-1)", ei_func->name);
		break;
	case -2: /* Thank you, come again */
		perr(0, "RTAS: %s: Busy, try again later (-2)", ei_func->name);
		break;
	case -3: /* Arg error */
		perr(0, "RTAS: %s: Argument error (-3)", ei_func->name);
		break;
	case -4: /* Call error */
		perr(0, "RTAS: %s: The error injection facility is not open\n"
			"or you are not the one that opened it", ei_func->name);
		break;
	case -1001: /* RTAS_KERNEL_INT */
		perr(0, "librtas: No Kernel Interface to Firmware");
		break;
	case -1002: /* RTAS_KERNEL_IMP */
		perr(0, "librtas: No Kernel Implementation of function %s",
		     ei_func->name);
		break;
	case -1003: /* RTAS_PERM */
		perr(0, "librtas: You must be root to access rtas calls");
		break;
	case -1004: /* RTAS_NO_MEM */
		perr(0, "librtas: Out of memory");
		break;
	case -1005: /* RTAS_NO_LOMEM */
		perr(0, "librtas: Kernel out of low memory");
		break;
	case -1006: /* RTAS_FREE_ERR */
		perr(0, "librtas: Attempt to free nonexistant rmo buffer");
		break;
	case -1007: /* RTAS_TIMEOUT */
		perr(0, "librtas: RTAS delay exceeded specified timeout");
		break;
	case -1098: /* RTAS_IO_ASSERT */
		perr(0, "librtas: %s: Unexpected I/O error", ei_func->name);
		break;
	case -1099: /* RTAS_UNKNOWN_OP */
		perr(0, "librtas: No firmware implementation of function %s",
		     ei_func->name);
		break;
	default:
		perr(0, "librtas returned an unknown error code (%d) for function %s",
		     rc, ei_func->name);
		break;
	}
}

/**
 * open_rtas_errinjct
 * @brief Open the RTAS error injection facility
 *
 * @param ei_func errinjct functionality
 * @return 0 on success, !0 otherwise
 */
int open_rtas_errinjct(ei_function *ei_func)
{
	int rc;

	rc = rtas_errinjct_open(&ei_token);
	if (rc != 0) {
		perr(0, "Could not open RTAS error injection facility");
		if (rc == -4)
			perr(0, "the facility is already open, please\n"
			     "specify the open token with the -k option");
		else if (rc == -5)
			perr(0, "PCI Error Injection is not enabled.");
		else
			check_librtas_returns(rc, ei_func);

	}

	return rc;
}

/**
 * close_rtas_errinjct
 * @brief Close the RTAS error injection facility
 *
 * @param ei_func errinjct functionality
 * @return 0 on success, !0 otherwise
 */
int close_rtas_errinjct(ei_function *ei_func)
{
	int rc;

	rc = rtas_errinjct_close(ei_token);
	if (rc != 0) {
		perr(0, "Could not close RTAS error injection facility");
		check_librtas_returns(rc, ei_func);
	}

	return rc;
}

/**
 * bind_cpu
 * @brief bind errinjct to a cpu
 *
 * Bind ourselves to a particular cpu if the cpu binding capability
 * is present on this machine.
 *
 * The inability to bind to a cpu is not considered a failure condition,
 * we simply print a message stating that cpu binding is not available
 * and return success.
 *
 * @return 0 on success, !0 otherwise
 */
static int bind_cpu(void)
{
	cpu_set_t mask;
	int rc;

	if (logical_cpu == -1)
		return 0;

	if (verbose)
		printf("Binding to logical cpu %d\n", logical_cpu);

	CPU_ZERO(&mask);
	CPU_SET(logical_cpu, &mask);

	rc = sched_setaffinity(getpid(), sizeof(mask), &mask);

	if (rc)
		perr(0, "Could not bind to logical cpu %d", logical_cpu);

	return rc;
}

/**
 * do_rtas_errinjct
 * @brief Perform the actual call to inject errors
 *
 * Perform the actual call to RTAS.  This routine will also bind us to
 * a specific processor (if requested) and open RTAS error injection if an
 * open token is not specified.
 *
 * @param ei_func errinjct functionality
 * @return 0 on success, !0 otherwise
 */
int do_rtas_errinjct(ei_function *ei_func)
{
	int rc = 0;
	int close_errinjct = 0;

	/* First see if we should bind ourselves to a processor */
	rc = bind_cpu();
	if (rc)
		return rc;

	/* Now, see if we need to open the RTAS errinjct facility */
	if (ei_token == -1) {
		rc = open_rtas_errinjct(ei_func);
		if (rc != 0)
			return rc;

		close_errinjct = 1;
	}

	/* Make the RTAS call */
	rc = rtas_errinjct(ei_func->rtas_token, ei_token, (char *)err_buf);
	if (rc != 0) {
		perr(0, "RTAS error injection failed!");
		check_librtas_returns(rc, ei_func);
		printf("This error may have occurred because error injection\n"
		       "is disabled for this partition. Please check the\n"
		       "FSP and ensure you have error injection enabled.\n");
	} else if (!be_quiet) {
		printf("Call to RTAS errinjct succeeded!\n\n");
	}

	if (close_errinjct)
		rc = close_rtas_errinjct(ei_func);

	return rc;
}

/**
 * read_ei_tokens
 * @brief Gather the RTAS tokens stored in /proc for this machine.
 *
 * This will read the RTAS token values stored in /proc to determine
 * which RTAS error injection capabilities are avaialable on this machine
 *
 * @return 0 on success, !0 otherwise
 */
int read_ei_tokens(void)
{
	char	buf[EI_BUFSZ];
	int	len;
	int	i, found;
	char	*tmp_ptr;
	int	fd;

	fd = open(EI_TOKEN_PROCFILE, O_RDONLY);
	if (fd == -1) {
		perr(errno, "Could not open %s", EI_TOKEN_PROCFILE);
		return 1;
	}

	len = read(fd, buf, EI_BUFSZ);
	if (len == -1) {
		perr(errno, "Could not read from %s", EI_TOKEN_PROCFILE);
		close(fd);
		return 1;
	}

	tmp_ptr = buf;
	while (tmp_ptr < buf + len) {
		found = 0;
		for (i = 0; i < NUM_ERRINJCT_FUNCS; i++) {
			if (strcmp(tmp_ptr, ei_funcs[i].name) == 0) {
				found = 1;
				tmp_ptr += strlen(tmp_ptr) + 1;
				ei_funcs[i].rtas_token = be32toh(*(int *)tmp_ptr);
				tmp_ptr += sizeof(int);
				break;
			}
		}

		if (found == 0) {
			if (verbose)
				perr(0, "Could not find errinjct function for rtas token \"%s\"",
				     tmp_ptr);
			tmp_ptr += strlen(tmp_ptr) + 1 + sizeof(int);
		}
	}

	return 0;
}

/**
 * sysfs_check
 * @brief Make sure sysfs is mounted at the expected /sys location
 *
 * Several places in errinjct need to look up data in sysfs.  In this
 * routine we make sure sysfs is mounted in the expected location /sys
 *
 * @return 0 if sysfs is mounted at /sys, !0 otherwise
 */
int sysfs_check(void)
{
	struct stat sbuf;
	int rc;

	rc = stat("/sys/class", &sbuf);

	/* posix semantics returns EOVERFLOW on sysfs in newer kernels */
	if (rc && errno == EOVERFLOW)
		rc = 0;

	if (rc)
		perr(errno, "It appears that sysfs is not mounted at /sys.\n"
		     "The error injection you requested requires sysfs,\n"
		     "please check your system configuration and try again.\n");

	return rc;
}


static struct option longopts[] = {
{
	name: "dry-run",
	has_arg: 0,
	flag: NULL,
	val: 254
},
{
	name: "help",
	has_arg: 0,
	flag: NULL,
	val: 'H'
},
{
	name: "verbose",
	has_arg: 0,
	flag: NULL,
	val: 'v'
},
{
	name: "quiet",
	has_arg: 0,
	flag: NULL,
	val: 'q'
},
{
	name: NULL
}
};

int main(int argc, char *argv[])
{
	ei_function *ei_func = NULL;
	const char *funcname;
	int	c, rc;
	int	i, fd;

	progname = argv[0];

	if (argc == 1) {
		ei_usage();
		exit(1);
	}

	if (get_platform() != PLATFORM_PSERIES_LPAR) {
		fprintf(stderr, "%s: is not supported on the %s platform\n",
							argv[0], platform_name);
		exit(1);
	}

	/* Make sure the error injection facility is available */
	fd = open(EI_IBM_ERRINJCT, O_RDONLY);
	if (fd == -1) {
		perr(0, "Could not open error injection facility,\n"
		     "file \"%s\" does not exist", EI_IBM_ERRINJCT);
		close(fd);
		exit(1);
	}

	close(fd);

	/* The function name is always first */
	funcname = argv[1];
	for (i = 0; i < NUM_ERRINJCT_FUNCS; i++) {
		if (ei_funcs[i].func == NULL)
			continue;

		if ((strcmp(funcname, ei_funcs[i].name) == 0)
		    || ((ei_funcs[i].alt_name != NULL)
		    && (strcmp(funcname, ei_funcs[i].alt_name)) == 0)) {
			ei_func = &ei_funcs[i];
			break;
		}
	}

	if (ei_func == NULL) {
		perr(0, "Could not find function \'%s\'", funcname);
		ei_ext_usage();
		exit(1);
	}

	/* shift past function name */
	argc--;
	argv++;

	while ((c = getopt_long(argc, argv, "+a:C:c:f:Hh:k:l:m:n:p:qs:v",
				longopts, NULL)) != -1) {
		switch (c) {
		case 254:
			dryrun = 1;
			break;
		case 'C':
			logical_cpu = atoi(optarg);
			break;
		case 'H':
			ext_help = 1;
			break;
		case 'k':
			ei_token = atoi(optarg);
			break;
		case 'q':
			be_quiet = 1;
			break;
		case 'v':
			/* If specified multiple times,
			 * turn on librtas debug
			 */
			if (verbose) {
				debug++;
				rtas_set_debug(debug);
			}
			verbose = 1;
			break;
		case ':':
			break;
		default:
			rc = ei_func->arg(c, optarg);
			if (rc) {
				perr(0, "\"-%c\" is not a valid option for %s",
				     c, ei_func->name);
				ext_help = 1;
			}
			break;
		}
	}

	/* Open the /proc/ppc64/rtas entry to get valid tokens */
	rc = read_ei_tokens();
	if (rc)
		exit(rc);

	memset(err_buf, 0, EI_BUFSZ);

	rc = ei_func->func(ei_func);

	exit(rc);
}
